src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegionIterator.java

```diff
@@ -0,0 +1,226 @@
+package org.broadinstitute.hellbender.engine;
+
+import htsjdk.samtools.SAMFileHeader;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.activityprofile.ActivityProfile;
+import org.broadinstitute.hellbender.utils.activityprofile.ActivityProfileState;
+import org.broadinstitute.hellbender.utils.activityprofile.BandPassActivityProfile;
+import org.broadinstitute.hellbender.utils.downsampling.DownsamplingMethod;
+import org.broadinstitute.hellbender.utils.iterators.AllLocusIterator;
+import org.broadinstitute.hellbender.utils.iterators.ReadCachingIterator;
+import org.broadinstitute.hellbender.utils.locusiterator.LocusIteratorByState;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadCoordinateComparator;
+import org.broadinstitute.hellbender.utils.read.ReadUtils;
+
+import java.util.*;
+
+/**
+ * Given a {@link Shard} of {@link GATKRead}, iterates over each {@link AssemblyRegion} within that shard, using the
+ * provided {@link AssemblyRegionEvaluator} to determine the boundaries between assembly regions.
+ *
+ * Loads the reads from the shard as lazily as possible to minimize memory usage.
+ *
+ * This iterator represents the core of the {@link AssemblyRegionWalker} traversal.
```

Comment:
Perhaps add to the contract of this method a warning about the needed unmapped reads filters to have been applied to the shard.

Reply:
Done